MINOR: Cleanup 'share' from group.coordinator.rebalance.protocols

elmoctarebnou opened the pull request:
Cleanup 'share' from group.coordinator.rebalance.protocols as a valid
value. Share Groups should instead be enabled through the share.version

Reviewers: Andrew Schofield <aschofield@confluent.io>, jimmy
 <wangzhiwang611@gmail.com>, TaiJuWu <tjwu1217@gmail.com>, Chia-Ping
 Tsai <chia7712@gmail.com>

elmoctarebnou committed:
Cleanup 'share' from group.coordinator.rebalance.protocols as a valid value. Share Groups should instead be enabled through the share.version

AndrewJSchofield commented on .gitignore:
```diff
@@ -31,6 +31,8 @@ TAGS
 Vagrantfile.local
 /logs
 .DS_Store
+.bloop/
```
Please revert this change.

elmoctarebnou replied:
Thanks for the review, @AndrewJSchofield. I’ve reverted the .gitignore change. Please let me know if there’s anything else I should adjust.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR.

elmoctarebnou committed:
Revert .gitignore change per review feedback

AndrewJSchofield commented:
@elmoctarebnou There are some build failures. Please check the logs and resolve. For example:

```
Error:  /home/runner/work/kafka/kafka/core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:1789:9: local var config in method testGroupCoordinatorRebalanceProtocols is never updated: consider using immutable val
```

elmoctarebnou committed:
Add explicit allowed list for group.coordinator.rebalance.protocols and test rejecting 'share'

elmoctarebnou commented:
Thanks, @AndrewJSchofield  — I fixed the error you pointed out. I also chatted with @clolov , and he suggested adding a small test to validate that "share" is not included in group.coordinator.rebalance.protocols, which I’ve now added.

elmoctarebnou commented:
Hi @AndrewJSchofield all tests are passing except for the ones already marked as flaky, which seem unrelated to this change.

JimmyWang6 commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1784,15 +1784,15 @@ class KafkaConfigTest {
     props.put(GroupCoordinatorConfig.GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, "consumer")
     assertThrows(classOf[ConfigException], () => KafkaConfig.fromProps(props))
 
+
```
nit: extra blank.

JimmyWang6 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java:
```diff
@@ -69,6 +69,11 @@ public class GroupCoordinatorConfig {
         Group.GroupType.CLASSIC.toString(),
         Group.GroupType.CONSUMER.toString(),
         Group.GroupType.STREAMS.toString());
+    public static final List<String> GROUP_COORDINATOR_REBALANCE_PROTOCOLS_ALLOWED = List.of(
```
I think we could reuse the `GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DEFAULT` constant since they are exactly the same.

elmoctarebnou replied:
Thanks @JimmyWang6 for the comment. I’ll update so ALLOWED just references DEFAULT.

JimmyWang6 reviewed (commented):
Thanks for the PR.

AndrewJSchofield commented:
Requesting review from @dajac since he's most knowledgeable about this config. I'm not certain the change is necessary or worthwhile given that this config is possibly being deprecated soon.

elmoctarebnou committed:
Reuse GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DEFAULT for ALLOWED

TaiJuWu commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1786,13 +1786,12 @@ class KafkaConfigTest {
 
-    // This is OK.
+    // Including "share" should be rejected.
     props.put(GroupCoordinatorConfig.GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, "classic,consumer,share")
```
Should we delete share here?

TaiJuWu replied:
Oh, I miss something, please ignore this comment.

chia7712 commented:
@elmoctarebnou could you please fix the conflicts?

elmoctarebnou committed:
Merge branch 'trunk' into remove-share-from-rebalance-protocols

elmoctarebnou commented:
Thank you @chia7712  for pointing out the conflict, somehow I missed this.

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java:
```diff
@@ -69,6 +69,7 @@ public class GroupCoordinatorConfig {
         Group.GroupType.CLASSIC.toString(),
         Group.GroupType.CONSUMER.toString(),
         Group.GroupType.STREAMS.toString());
+    public static final List<String> GROUP_COORDINATOR_REBALANCE_PROTOCOLS_ALLOWED = GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DEFAULT;
```
why not using `GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DEFAULT` directly?

elmoctarebnou replied:
@chia7712 thank you for the comment! The `GROUP_COORDINATOR_REBALANCE_PROTOCOLS_ALLOWED` is just for readability.

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java:
```diff
@@ -298,7 +299,8 @@ public class GroupCoordinatorConfig {
         // Group coordinator configs
         .define(GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, LIST, GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DEFAULT,
-                ConfigDef.ValidList.in(false, Group.GroupType.documentValidValues()), MEDIUM, GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DOC)
+                ConfigDef.ValidList.in(GROUP_COORDINATOR_REBALANCE_PROTOCOLS_ALLOWED.toArray(new String[0])), 
```
not sure why `share` is not controlled by `group.coordinator.rebalance.protocols`? By contrast, `stream` and `consumer` could be disabled by `group.coordinator.rebalance.protocols`

AndrewJSchofield replied:
That's because share groups are controlled by a feature. This config was created before the features were implemented to turn on KIP-848 and KIP-932. I understand the plan is to deprecate and remove it entirely. It's a really bad idea having a broker-level config that controls this because individual brokers might be misconfigured and a cluster is in a mixed state. That's why a feature is so much better.

chia7712 replied:
@AndrewJSchofield Thanks for the response. Do we have a KIP that covers deprecating this config? Since each protocol already maintains its own feature version, it feels like the right time to deprecate it. Otherwise, exposing an incomplete public config seems  a bit odd to me

AndrewJSchofield replied:
No KIP yet. But then KIP-932 says that share groups are enabled by a feature, and not `group.coordinator.rebalance.protocols`.

AndrewJSchofield replied:
@chia7712 I do understand your point but I was not expecting other group types to use this config after KIP-848. It would be acceptable to me for share groups to honour this config in trunk and AK 4.2. Let me know what you think.

chia7712 replied:
> I do understand your point but I was not expecting other group types to use this config after KIP-848. It would be acceptable to me for share groups to honour this config in trunk and AK 4.2. Let me know what you think.

In fact, I’d prefer to raise a KIP to deprecate group.coordinator.rebalance.protocols in trunk and AK 4.2. Let’s leave it as is, since extending a deprecated config would be a bit odd. I’ve opened KAFKA-19740 for this.

AndrewJSchofield replied:
@dajac I think you're planning a KIP to complete the new consumer group protocol migration, so I expect this is on your radar.

elmoctarebnou commented:
Related PR: [apache/kafka-site#717](https://github.com/apache/kafka-site/pull/717)

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java:
```diff
@@ -298,7 +299,8 @@ public class GroupCoordinatorConfig {
     public static final ConfigDef CONFIG_DEF = new ConfigDef()
         // Group coordinator configs
         .define(GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, LIST, GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DEFAULT,
-                ConfigDef.ValidList.in(false, Group.GroupType.documentValidValues()), MEDIUM, GROUP_COORDINATOR_REBALANCE_PROTOCOLS_DOC)
```
It should be `ConfigDef.ValidList.in(false, GROUP_COORDINATOR_REBALANCE_PROTOCOLS_ALLOWED.toArray(new String[0]))`, right? The empty `GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG` is illegal

elmoctarebnou replied:
@chia7712 You're right — thanks for catching that! I’ve updated the validation to disallow empty lists

elmoctarebnou committed:
Disallow empty group.coordinator.rebalance.protocols

chia7712 reviewed (approved):
LGTM

chia7712 commented:
@AndrewJSchofield What do you think about this PR?

AndrewJSchofield commented:
> @AndrewJSchofield What do you think about this PR?

It's fine. I slightly prefer to permit and ignore `"share"` in the config, but I know it's just a matter of taste.

chia7712 commented:
> I slightly prefer to permit and ignore "share" in the config, but I know it's just a matter of taste.

This seems to be related to a compatibility issue. The config `group.coordinator.rebalance.protocols=classic,share` should not throw an exception, even though `share` is a no-op

@elmoctarebnou what's your take on this?

elmoctarebnou commented:
Thanks @chia7712 and @AndrewJSchofield for the feedback — I agree that maintaining compatibility makes sense here. I’ll update the PR so that "share" is accepted in `group.coordinator.rebalance.protocols` but treated as a no-op (ignored internally with a warning). Let me know if you’re good with that approach, or if you’d prefer a different handling.

chia7712 commented:
> I’ll update the PR so that "share" is accepted in group.coordinator.rebalance.protocols but treated as a no-op (ignored internally with a warning)

sounds good to me :smiley:

AndrewJSchofield commented:
> I’ll update the PR so that "share" is accepted in `group.coordinator.rebalance.protocols` but treated as a no-op (ignored internally with a warning). Let me know if you’re good with that approach, or if you’d prefer a different handling.

Works for me too. Thanks.

elmoctarebnou committed:
Accept 'share' in group.coordinator.rebalance.protocols with a warning log

chia7712 commented:
@elmoctarebnou Could you please merge `trunk` to bring in the recent CI updates?

elmoctarebnou committed:
Merge remote-tracking branch 'upstream/trunk' into remove-share-from-rebalance-protocols

AndrewJSchofield reviewed (approved):
Looks good to me.
